drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1532 @@
+	// flush if any left
+	if (item) {
+		// always pass first item
+		item = bdata.item_refs[bij.first_item_ref].m_pItem;
 
-	glDisable(GL_BLEND);
+		flush_render_batches(item, current_clip, reclip, p_material);
+	}
+}
 
-	Vector2 wh(storage->frame.current_rt->width, storage->frame.current_rt->height);
+void RasterizerCanvasGLES2::flush_render_batches(Item *p_item, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
+	//print_line("_flush_render_batches");
 
-	Color copy_section(p_rect.position.x / wh.x, p_rect.position.y / wh.y, p_rect.size.x / wh.x, p_rect.size.y / wh.y);
+	// some heuristic to decide whether to use colored verts.
+	// feel free to tweak this.
+	// this could use hysteresis, to prevent jumping between methods
+	// .. however probably not necessary
+	if ((bdata.total_color_changes * 4) > (bdata.total_quads * 3)) {
+		bdata.use_colored_vertices = true;
 
-	if (p_rect != Rect2()) {
-		storage->shaders.copy.set_conditional(CopyShaderGLES2::USE_COPY_SECTION, true);
+		// small perf cost versus going straight to colored verts (maybe around 10%)
+		// however more straightforward
+		_batch_translate_to_colored();
+	} else {
+		bdata.use_colored_vertices = false;
 	}
 
-	storage->shaders.copy.set_conditional(CopyShaderGLES2::USE_NO_ALPHA, !state.using_transparent_rt);
+	// send buffers to opengl
+	_batch_upload_buffers();
 
-	glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->copy_screen_effect.fbo);
-	glActiveTexture(GL_TEXTURE0);
-	glBindTexture(GL_TEXTURE_2D, storage->frame.current_rt->color);
+	Item::Command *const *commands = p_item->commands.ptr();
+	//int s = p_item->commands.size();
 
-	storage->shaders.copy.bind();
-	storage->shaders.copy.set_uniform(CopyShaderGLES2::COPY_SECTION, copy_section);
+	render_batches(commands, 0, current_clip, reclip, p_material);
+}
 
-	const Vector2 vertpos[4] = {
-		Vector2(-1, -1),
-		Vector2(-1, 1),
-		Vector2(1, 1),
-		Vector2(1, -1),
-	};
+void RasterizerCanvasGLES2::_canvas_item_render_commands(Item *p_item, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
 
-	const Vector2 uvpos[4] = {
-		Vector2(0, 0),
-		Vector2(0, 1),
-		Vector2(1, 1),
-		Vector2(1, 0)
-	};
+	int command_count = p_item->commands.size();
+	int command_start = 0;
 
-	const int indexpos[6] = {
-		0, 1, 2,
-		2, 3, 0
-	};
+	Item::Command *const *commands = p_item->commands.ptr();
 
-	_draw_polygon(indexpos, 6, 4, vertpos, uvpos, NULL, false);
+	// while there are still more batches to fill...
+	// we may have to do this multiple times because there is a limit of 65535
+	// verts referenced in the index buffer (each potential run of this loop)
+	while (command_start < command_count) {
+		{
+			// legacy .. just create one massive batch and render everything as before
+			bdata.batches.reset();
```

Comment:
This section confuses me. It appears to be a while loop that only even runs once. Additionally, it creates a default batch that is never used. Is this just from a copy-paste from the command batching PR?

Reply:
Ah well spotted, this may be a historical artefact. I'll see if it can be simplified.